### PR TITLE
#174 script to retry failed webhooks

### PR DESCRIPTION
**Pull request by GauravGusain98**
#174 

### Types of changes

- [ ] Bug fix (a change which fixes an issue)
- [x] New feature (a change which adds functionality)

### Short description of what this resolves:

- Added functionality to retry failed webhook logs 
- The created on a date is set to the created on the date of logs.

### Checklist:

- [x] I have performed a self-review of my own code.
- [x] The code follows the style guidelines of this project.
- [x] The code changes are passing the CI checks
- [ ] I have documented my code wherever required.
- [ ] The changes requires a change to the documentation.
- [ ] I have updated the documentation based on the my changes.

### CODE REVIEW

File: api/services/call_log_service.py [hunk 1]
```diff
@@ -96,6 +96,9 @@ def create_call_logs(self, jsonData):
                 flow_category=jsonData["flow_category"]
                 if "flow_category" in jsonData
                 else self.flow_category,
+                created_on=jsonData["log_created_on"]
+                if jsonData.get("log_created_on", None)
+                else datetime.now(),
```
Comment (GauravGusain98) [inline api/services/call_log_service.py:101]: We will only receive `jsonData["log_created_on"]` for the retry logic.

File: api/services/transaction_log_service.py [hunk 2]
```diff
@@ -14,3 +14,12 @@ def create_new_webhook_log(self, jsonData):
     def mark_webhook_log_as_processed(self, webhook_log):
         webhook_log.processed = True
         helpers.save(webhook_log)
+
+    def get_failed_webhook_transaction_log(self):
+        failed_webhook_transaction_logs = (
+            models.WebhookTransactionLog.query.filter_by(processed=False)
+            .limit(app.config["RETRY_LOGS_LIMIT"])
```
Comment (GauravGusain98) [inline api/services/transaction_log_service.py:21]: taking limit of the failed logs count using env

File: main.py [hunk 3]
```diff
@@ -1,67 +1,103 @@
...
-
-                # Handle custom fields
-                if (
-                    "fields" in jsonData["contact"]
-                    and jsonData["contact"]["fields"] is not None
-                    and jsonData.get("flow_category", None) == "dry_flow"
-                ):
-                    handle_user_custom_field_data(jsonData)
+                webhook_log = transaction_log_service.create_new_webhook_log(jsonData)
+            processed = handle_payload(jsonData)
```
Comment (GauravGusain98) [inline main.py:20]: Inside handle_payload function try and except are used for error handling and based on that it returns True or False

File: main.py [hunk 4]
```diff
@@ -1,67 +1,103 @@
...
+                return jsonify(message="Something went wrong!"), 400
+            elif processed == -1:
                 return jsonify(message="Contact"), 400
-        except:
-            return jsonify(message="Something went wrong!"), 400
+
+            if "contact" in jsonData:
+                transaction_log_service.mark_webhook_log_as_processed(webhook_log)
+        elif jsonData and jsonData.get("type", None) == "retry_failed_log":
+            retry_failed_webhook(transaction_log_service)
```
Comment (GauravGusain98) [inline main.py:30]: Call retry login if we have `"type": "retry_failed_log"` in json data.

File: main.py [hunk 5]
```diff
@@ -1,67 +1,103 @@
...
         return jsonify(message="Currently, the system do not accept a GET request"), 405
 
 
+def retry_failed_webhook(transaction_log_service):
+    failed_webhook_logs = transaction_log_service.get_failed_webhook_transaction_log()
+
+    for log in failed_webhook_logs:
+        json_data = json.loads(log.payload)
+        json_data["log_created_on"] = log.created_on
+        processed = handle_payload(json_data, True)
```
Comment (GauravGusain98) [inline main.py:43]: Processing all the failed logs. The second parameter is for differentiating regular webhooks and retry webhooks

**Review by Satendra-SR — APPROVED** [review]
Just one suggestion else looks good. Please fix and merge

**Pull request merged**